### PR TITLE
DR-1063 Fix handling of GCS paths containing '#'

### PR DESCRIPTION
**Pull request by danxmoran**
For example, the HCA's v1 data store contains files named like:
```
042b4e00-82f4-47a5-9e82-408cd082fe26_2019-05-16T032622.030624Z_21935_6#96_2.fastq.gz
```
The existing URI parsing was chopping off everything after the `#`

### CODE REVIEW

File: src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java
Comment (danxmoran) [inline src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java]: Not sure if it's worth having this check here. I copied most of this logic from `IngestUtils.parseBlobUri`, which encodes the check in a slightly different way. It'd probably be good to move the parsing into a single spot that both of these can use.

File: src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java]: I think this should be an uppercase name as a constant.

File: src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java
Comment (ddietterich) [reply src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java]: I made [DR-1066](https://broadworkbench.atlassian.net/browse/DR-1066) for this tech debt.

File: src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java
Comment (ddietterich) [inline src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:189]: The old code had some bucket name validation. Should we have some check that makes sure the bucket name at least obeys the GCS naming rules? Or catch an error from the getBlob?

**Review by ddietterich — COMMENTED** [review]
Minor comments

**Review by fboulnois — APPROVED** [review]
lgtm, I agree with DD's earlier comment regarding bucket name validation.

**Review by thathert — APPROVED** [review]
😞 This goes against the URL spec -- `#` should be used for fragment identifiers. Seems too late to change this now, so approving assuming the tests look okay.

File: src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:189]: :+1: will add more checks

**Comment by danxmoran** [comment]
> 😞 This goes against the URL spec -- # should be used for fragment identifiers. Seems too late to change this now, so approving assuming the tests look okay.

FWIW the Google SDKs do URL-encode object names before hitting the base APIs. I'd be wary of any codepath that might (purposefully or accidentally) pre-encode pieces of the path, since it could cause double-encoding (and I'm not sure if the [GCS encoding scheme](https://cloud.google.com/storage/docs/request-endpoints#encoding) aligns with the official URL spec).

File: src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java
Comment (danxmoran) [reply src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java:189]: Covered everything I could see from the GCS docs

File: src/test/java/bio/terra/integration/FileTest.java
Comment (danxmoran) [inline src/test/java/bio/terra/integration/FileTest.java:123]: This is a change in existing behavior. I would advocate that it's a good thing, because it requires less thinking on the user's part and aligns with gsutil:
```
$ gsutil stat 'gs://jade-testdata/files/File%20Design%20Notes.pdf'
No URLs matched: gs://jade-testdata/files/File%20Design%20Notes.pdf

$ gsutil stat 'gs://jade-testdata/files/File Design Notes.pdf'
    Creation time:          Tue, 16 Apr 2019 14:00:01 GMT
    Update time:            Tue, 16 Apr 2019 14:00:01 GMT
    Storage class:          REGIONAL
    Content-Length:         301354
    Content-Type:           application/pdf
    Hash (crc32c):          Y5gazg==
    Hash (md5):             6n/gNRj/0F/OwjR5YwMJiA==
    ETag:                   CIHAifvi1OECEAE=
    Generation:             1555423201222657
    Metageneration:         1
```
But worth calling out.

Comment (ddietterich) [reply src/test/java/bio/terra/integration/FileTest.java:123]: Yes, it is a good thing. Thanks!

**Review by ddietterich — APPROVED** [review]
LGTM - thanks!

**Comment by danxmoran** [comment]
Got a +1 from Jeremy to merge even though integration is failing. Thanks for the quick turn-around everybody!

**Pull request merged**
